### PR TITLE
SNO-174-add-local-storage-redis

### PR DESCRIPTION
**Pull request by caseylitton**
SNO-175 is on a different PR but included for testing reasons

### CODE REVIEW

File: src/snovault/local_storage.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    LOCAL_STORAGE_EVENT_TAGS,
+    LOCAL_STORAGE_STATE,
+)
+
+
+def includeme(config):
+    config.scan(__name__)
+    config.add_route('local_storage', '/local_storage')
+    config.registry[LOCAL_STORAGE] = LocalStoreClient(
+        db=config.registry.settings['local_storage_db'],
```
Comment (pedrohr) [inline src/snovault/local_storage.py]: this is correctly set as a registry setting, however, I wonder if there's any hidden pyramid magic that can give you the application name (main, indexer, regionindexer, etc...) and then have a hash map making the translation to the DB index number. If there isn't, maybe we could rename the variable to `local_storage_db_redis_index` or something like that, people might get lost when they see that for the first time. Or add a comment in the .ini files (maybe the comment is easier :))

File: src/snovault/local_storage.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        'utc_now': str(datetime.utcnow()),
+        'lcl_now': f"{local_tz}: {datetime.now(timezone(local_tz))}",
+        # Different pyramid process may have different registries, this returns memory address
+        'local_storage': str(local_store),
+    }
+
+    # Set/Get dict or hash
+    redis_info = local_store.dict_get(LOCAL_STORAGE_STATE)
+    if not redis_info:
+        redis_info = {
```
Comment (pedrohr) [inline src/snovault/local_storage.py]: should this be part of the LocalStoreClient initializer, so we always have a state ready?

File: src/snovault/local_storage.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    # Set/Get single item
+    result['counter'] = int(local_store.item_get('counter')) + 1
+    local_store.item_set('counter', result['counter'])
+
+    # Store event key/value pairs with tag and list
+    event_key = request.params.get("event_key")
+    event_value = request.params.get("event_value")
+    event_tag = request.params.get("event_tag")
+    if event_key and event_value:
+        # Key values pairs are tagged for easier retrieval/organization
```
Comment (pedrohr) [inline src/snovault/local_storage.py]: I think we should split this endpoint into two:
GET /events => get event info by tag or all events
POST /events => create a new event from a pair (event_key, event_value)
or any similar name for the endpoints :)

File: src/snovault/local_storage.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        if not event_tag:
+            # New tag created
+            event_tag = local_store.get_tag(4)
```
Comment (pedrohr) [inline src/snovault/local_storage.py]: I would add a comment on the `4`. Bitwise operations can get tricky

File: src/snovault/local_storage.py (hunk 1 above)
Comment (caseylitton) [reply src/snovault/local_storage.py]: yea, the naming could be better.  There are a few others I might fix along with this one.

The only way I know to get the name of the section in the config is to add a var called name to the section in the config :)    We sorta do this in the ini files for encoded with the set indexer = true, set visindexer = true...at one point I named them explicitly.  Must have been in a feature that didn't get added.

File: src/snovault/local_storage.py (hunk 2 above)
Comment (caseylitton) [reply src/snovault/local_storage.py]: That's a good idea.  It's more static that way.  Also on initialization the connection will be checked.  Connection errors don't happen until a query is made.

File: src/snovault/local_storage.py (hunk 1 above)
Comment (caseylitton) [reply src/snovault/local_storage.py]: oh, and yea.  I think the db indexes should be global to all processes.  How would we get indexer data in the app if the app didn't know the indexer db index.

File: src/snovault/local_storage.py (hunk 3 above)
Comment (caseylitton) [reply src/snovault/local_storage.py]: these were moved SNO-173, this PR doesn't have an implementation of the local store.  It seemed like unnecessary code.

**Comment by keenangraham** [comment]
Note to rebase this on dev before merge.

**Review by pedrohr — APPROVED** [review]
💯

**Pull request merged**